I538 support interactive problems

johnbrvc opened the pull request:
### Description of what the PR does
Add built-in support for interactive problems.
Previously, interactive problems were handled by munging the execute command line from the edit problem GUI or in a problem's YAML file.
This PR makes PC2 aware of interactive problems.  Special custom shell scripts are used to setup the pipe work needed to connect a run's I/O to the O/I of a CLICS compliant custom interactive validator.  There are separate scripts for sandbox and non-sandbox problems.
An additional special script is used during the PC2 `validateProgram()` phase that copies the results of an interactive run performed by one of the aforementioned shell scripts to the location desired by PC2.

Summary of changes to the code:

- Add support in `ContestSnakeYAMLLoader.java` to recognize an interactive problem and perform appropriate validation of the validator supplied.
- Add support to the `ProblemPane` GUI to indicate if a problem is interactive.
- Add support to the `EditProblemPane `GUI to allow configuration of interactive problems.
- Add support for interactive problems to the `Problem` object.
- Expandthe `CustomValidatorSettings `object to be aware of interactive problems.
- `Executable.java`

  - Change to deal with interactive problems.
  - Refactor some code to put commonly used blocks in methods.
  - Add more substitution variables to support interactive problems and also allow for better logging in validators (`{:testcase}`, `{:executeinfofilenam}`) .
  - Add routines to set up the interactive environment in the execute folder.

- Create new scripts folder to hold all helper scripts used by sandboxes and interactive validators.  This involved getting rid of the `sandbox `folder (really, renaming it) and creating the `scripts `folder. 
- CI: Fix missing "`toString`" properties.
- CI: Make` teams2.tsv` be the preferred teams file during loading of a CDP.  This provides organizations.
- CI: better reporting of errors during reading of the CDP when loading a contest.  Combats the antagonizing "_check logs_" message previously reported.
- CI: Update copyright notices

### Issue which the PR addresses
Fixes #538

### Environment in which the PR was developed (OS,IDE, Java version, etc.)
Windows 11/Ubuntu 22.04.2
Eclipse
java version "1.8.0_321"
Java(TM) SE Runtime Environment (build 1.8.0_321-b07)
Java HotSpot(TM) 64-Bit Server VM (build 25.321-b07, mixed mode)

### Precise steps for _testing_ the PR 
Coming soon.

johnbrvc committed:
Add support scripts for interactive problems

These shell scripts are called from PC2 to implement interactive problem judging.

johnbrvc committed:
Add support for validation property

If a problem's yaml file has a "validation" property with the value "custom interactive", then make sure the necessary validator is supplied and meets the CLICS criteria.
Add a flag in Problem to remember if the problem is interactive or not based on the value of the "validation" property.

johnbrvc committed:
Add basic interactive problem support

Added constants for scripts to call for interactive problem.
Added constant for submission "executeinfo" json log which is used to provide detailed information about each test case run for a submission.
Minimal changes to Executable to support calling the interactive scripts for interactive problems only (problem.isInteractive()).
Add {:testcase} replacement variable indicating test case #.
Add {:executeinfofilename} containing the ndjson file to send per test case information to.  More correctly implement sandbox command/program strings (pay attention to the sandbox type).
Add string for interactiveCommandLine (currently, this is constant).

johnbrvc committed:
Merge branch 'develop' into i538_support_interactive_problems

johnbrvc committed:
Copyright date update

johnbrvc committed:
Correctly handle missing teams.tsv if teams2.tsv is present

If teams2.tsv was present, but teams.tsv was not, pc2 would not load the CDP, indicating that teams.tsv was missing.  teams.tsv is NOT required if teams2.tsv is present.
Added comment to checkFiles() routine to describe what the routine does.
Removed extra / duplicate code that was already done below, and, shouldn't have been there in the first place.

johnbrvc committed:
CI: Give better GUI error message on cdp load failure

Previously, if an exception was thrown during a CDP load, a message would appear indicating some error occurred and you should check the logs.
Now, the GUI dialog box indicates what the specific error was (saves a trip to the logs for easy stuff like files not found, etc).

johnbrvc committed:
Fixes to support loading teams2.tsv

If only a teams2.tsv is present, the cdp would not load.  This has been fixed.
Wrong variable was used to check for custom validator.

johnbrvc committed:
Move interactive flag to CustomValidatorSettings

The interactive flag is really part of custom validation.  Move the notion of interactive problem to the CustomValidatorSettings object.
Change Snake yaml reader to set the flag in the CustomValidatorSettings object instead of the Problem.

johnbrvc committed:
Add GUI for setting interactive validator

Display Interactive for Input Type field.
Add additional type of custom output validator, Interactive.

johnbrvc committed:
Merge branch 'develop' into i538_support_interactive_problems

johnbrvc committed:
Move all interactive/sandbox scripts to a single folder

Due to the increasing number of special scripts to run sandboxes and interactive validators, scripts were moved to a single folder PC2HOME/scripts, instead of having separate folders for sandbox and interactive.
Changed code to use Constant.PC2_SCRIPT_DIRECTORY when using scripts.
Fixed missing space in interactive sandbox validator command line.
Indicate whether a problem uses Interactive I/O on the ProblemsPane.

johnbrvc committed:
Merge branch 'develop' into i538_support_interactive_problems

johnbrvc committed:
Get full paths of judges data and validator

Since the full paths of judge data and the interactive validator are needed, we created methods that make these up and other places can use them. (Common code).

johnbrvc committed:
Make scripts have better reporting.  Bug fixes.

There were some issues with ordinal vs. index of test cases.  Fix those.
Fix scripts so they are more robust and report errors better.
Fix comments indicating additional substitution variables.
Force use of pc2 validator for interactive problems at validation phase.

johnbrvc committed:
Merge branch 'develop' into i538_support_interactive_problems

johnbrvc committed:
Cleanup comments

lane55 commented on scripts/pc2_interactive.sh:
```diff
@@ -0,0 +1,311 @@
+interactive-validator, judge's program to execute to provide interactive responses to the submission
+judgein, judges input file
+judgeans, judges answer file
+testcase, test case number (for logging)
```
missing command and command_args descriptions

johnbrvc replied:
Fixed.

lane55 commented on scripts/pc2_interactive.sh:
```diff
@@ -0,0 +1,311 @@
+# ------------------------------------------------------------
+
+if [ "$#" -lt 1 ] ; then
+   echo $0: No command line arguments
```
Missing commnd line...

lane55 commented on scripts/pc2sandbox_interactive.sh:
```diff
@@ -0,0 +1,448 @@
+interactive-validator, judge's program to execute to provide interactive responses to the submission
+judgein, judges input file
+judgeans, judges answer file
+testcase, test case number (for logging)
```
Add command and command_args description

johnbrvc replied:
Fixed.

lane55 commented on src/edu/csus/ecs/pc2/core/Constants.java:
```diff
@@ -188,13 +188,40 @@ private Constants() {
+    /**
+     * Constants for interactive problems
+     */
+    // Filename of the script to run for an interactive problem in a sandbox 
```
Incorrect comment layout, for javadoc and quick lookup in eclipse, use layout
```
/**
 * <describe field>
 */
```

johnbrvc replied:
Fixed.

lane55 reviewed (approved):
Minor comments

johnbrvc committed:
Better help messages in scripts.

Added description of command and command arguments to interactive scripts.
Changed wording of error when no arguments are supplied to make it more user friendly.

johnbrvc committed:
Fix javadoc for sandbox and interactive problem constants

Change // comments to javadoc comments.

johnbrvc committed:
Clean up TODO's

For TODO's that are for future development, label them as such.
Introduce new debug variable for allowing the debugging of interactive problem execution flow on Windows.

johnbrvc committed:
Cleanup the checking for team and group TSV files

Fix comment to it better explains the changes made.
Change code so it's more compact, and hopefully, more readable.

johnbrvc committed:
Clarify TODO

Push this TODO for moving messages out of source code to future development.

johnbrvc committed:
Refactor newly added enum

The JudgeFileType enum already exists in the Utilities class by a different name, so use that.
Refactor code to make use of new getJudgeFileName() method.
If an error occurs when looking up a judge's file, show the expected name of the file, if it's available.  This will aid debugging.
Amend javadoc comment in Utilities.locateJudgesDataFile() to indicate that the last argument is never used, since the serialized file is already passed in.

johnbrvc committed:
Add sample contest for interactive

Allows testing of interactive problems along with a non-interactive problem for completeness.

johnbrvc committed:
Fixed bug with setting up interactive scripts

Scripts have to be setup for interactive only if problem is in fact interactive.  This bug fixes a problem where if the problem wasnt interactive, but had a sandbox, it would try to copy the interactive validator and fail.

johnbrvc committed:
Fixed bug in accepted solution for guess interactive problem

The accepted solution had a bug.  It produced a floating point value with a decimal point when the problem statement explicitly says integer guesses only.  Copied version with the bug to the wrong_answer folder (guess_nonint.py), and fixed the bug in the accepted soltuion.

johnbrvc committed:
Remove unused imports

johnbrvc committed:
Merge branch 'develop' into i538_support_interactive_problems

jbuck-cs committed:
Fixed modes on files.

johnbrvc committed:
Merge branch 'i538_support_interactive_problems' of https://github.com/johnbrvc/pc2v9 into i538_support_interactive_problems

jbuck-cs committed:
Default pin to CPU 1 (since everyone has at least one cpu)
Copy sample data for the non-interactive problem to the secret folder for the
case where the load sample PR has not been merged yet.

johnbrvc committed:
Merge branch 'i538_support_interactive_problems' of https://github.com/johnbrvc/pc2v9 into i538_support_interactive_problems

johnbrvc committed:
Update package build scripts

Add the new "scripts" folder to the package build, really, renaming the old "sandbox" folder and making it more generic.

johnbrvc committed:
Update package build script to include scripts folder

The sandbox folder has been renamed scripts, since there are now other scripts unrelated to sandboxes.  It makes sense to put them all in one place.  These do not belong in the bin folder since they are internal only to pc2.

johnbrvc committed:
Merge branch 'develop' into i538_support_interactive_problems

jbuck-cs committed:
Fixed path bug in script.  We want to execute validator from the execute folder.

johnbrvc committed:
i538 Fix for comments in PR about validation type checking

The code did not confirm that the string passed in the "validation" property for a problem was valid.  It only looked for strings it cared about.  Anything else it simply ignored.  This fix makes the code more robust and complete.

johnbrvc committed:
Merge branch 'develop' into i538_support_interactive_problems

jbuck-cs committed:
i538 Fixed interactive validators to ignore SIGPIPE

The interactive validators would get SIGPIPE if the submission went away before the IV sent its output.  This prevents that from happening.

johnbrvc committed:
Merge branch 'develop' into i538_support_interactive_problems

johnbrvc committed:
i530 Changes requested in PR to comments and nomenclature

Addresses several comments made in the PR798 discussion.

jbuck-cs committed:
Changed comment to reflect correct name of script.
